pkg/ext/store.go

```diff
@@ -2,92 +2,260 @@ package ext
+) (runtime.Object, bool, error) {
+	oldObj, err := getFn(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, false, err
+		}
+
+		obj, err := objInfo.UpdatedObject(ctx, nil)
+		if err != nil {
+			return nil, false, err
+		}
+
+		if err = createValidation(ctx, obj); err != nil {
+			return nil, false, err
+		}
+
+		tObj, ok := obj.(T)
+		if !ok {
+			var zeroT T
+			return nil, false, fmt.Errorf("object was of type %T, not of expected type %T", obj, zeroT)
+		}
+
+		newObj, err := createFn(ctx, tObj, &metav1.CreateOptions{})
+		if err != nil {
+			return nil, false, err
+		}
+		return newObj, true, err
+	}
+
+	newObj, err := objInfo.UpdatedObject(ctx, oldObj)
+	if err != nil {
+		return nil, false, err
+	}
+
+	newT, ok := newObj.(T)
+	if !ok {
+		var zeroT T
+		return nil, false, fmt.Errorf("object was of type %T, not of expected type %T", newObj, zeroT)
+	}
+
+	if updateValidation != nil {
+		err = updateValidation(ctx, newT, oldObj)
+		if err != nil {
+			return nil, false, err
+		}
+	}
+
+	newT, err = updateFn(ctx, newT, options)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return newT, false, nil
+}
+
+type watcher struct {
+	closedLock sync.RWMutex
+	closed     bool
+	ch         chan watch.Event
+}
+
+func (w *watcher) Stop() {
+	w.closedLock.Lock()
+	defer w.closedLock.Unlock()
+	if !w.closed {
+		close(w.ch)
+		w.closed = true
+	}
+}
+
+func (w *watcher) addEvent(event watch.Event) bool {
+	w.closedLock.RLock()
+	defer w.closedLock.RUnlock()
+	if w.closed {
+		return false
+	}
+
+	w.ch <- event
+	return true
+}
+
+func (w *watcher) ResultChan() <-chan watch.Event {
+	return w.ch
 }
 
 type WatchEvent[T runtime.Object] struct {
 	Event  watch.EventType
 	Object T
 }
+
+func doWatch[T runtime.Object](
+	ctx context.Context,
+	scheme *runtime.Scheme,
+	internaloptions *metainternalversion.ListOptions,
+	watchFn func(context.Context, *metav1.ListOptions) (chan WatchEvent[T], error),
+) (watch.Interface, error) {
+	options, err := convertListOptions(internaloptions, scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &watcher{
+		ch: make(chan watch.Event),
+	}
+	go func() {
+		// Not much point continuing the watch if the store stopped its watch.
+		// Double stopping here is fine.
+		defer w.Stop()
+
+		// Closing eventCh is the responsibility of the store.Watch method
+		// to avoid the store panicking while trying to send to a close channel
+		eventCh, err := watchFn(ctx, options)
+		if err != nil {
+			return
+		}
+
+		for event := range eventCh {
+			added := w.addEvent(watch.Event{
+				Type:   event.Event,
+				Object: event.Object,
+			})
+			if !added {
+				break
+			}
+		}
+	}()
+
+	return w, nil
+}
+
+func convertListOptions(options *metainternalversion.ListOptions, scheme *runtime.Scheme) (*metav1.ListOptions, error) {
+	var out metav1.ListOptions
+	err := scheme.Convert(options, &out, nil)
+	if err != nil {
+		return nil, fmt.Errorf("convert list options: %w", err)
```

Comment:
```suggestion
		return nil, fmt.Errorf("error converting list options: %w", err)
```

Reply:
I usually prefer not to prefix every error with `error`, it's usually clear that we're getting an error.

eg: I prefer to avoid this

```
error converting list options: error this didn't work: error foo failed
```